### PR TITLE
Use Map instead of Object for RuleSets.targets

### PR DESCRIPTION
**Pull request by cschanaj**
@Hainish this will notably improve performance and memory usage, as suggested in https://github.com/EFForg/https-everywhere/issues/1775#issuecomment-151655897, whereas the performance is measured by https://jsperf.com/es6-map-vs-object-properties/42. 

~Why `Firefox` test always fails?~

### CODE REVIEW

File: chromium/rules.js [hunk 1]
```diff
@@ -328,9 +333,9 @@ RuleSets.prototype = {
 
     var tmp;
```
Comment (cschanaj) [inline chromium/rules.js]: BTW, this variable is not being used. Should I remove it in this PR?

Comment (koops76) [reply chromium/rules.js]: @cschanaj I will remove unused variables in one of my PRs.

**Comment by Hainish** [comment]
Checking on the `firefox-dev` test now.

**Comment by Hainish** [comment]
Despite a [note to the contrary](https://wiki.mozilla.org/Add-ons/Extension_Signing#FAQ), it looks like the developer edition is now strictly enforcing xpi signatures.  I've updated the https://github.com/EFForg/https-everywhere-docker-base URL to point to the unbranded edition of Firefox Dev, so once that is done building I can re-run the tests and we should be good.

**Comment by koops76** [comment]
@Hainish Nightly is unbranded by default and does not enforce extension signing.

**Comment by koops76** [comment]
@Hainish Nightly is unbranded by default and does not enforce extension signing.

**Comment by cschanaj** [comment]
> To disable signature checks, you will need to set the xpinstall.signatures.required preference to "false". 

@Hainish It seems that you'd have to modify `test/script.py` to disable signature checking in the Developer Edition. Did you test it?

**Comment by Hainish** [comment]
@cschanaj this is already set by https://github.com/EFForg/https-everywhere/blob/master/test/firefox/test_profile_skeleton/prefs.js#L10

**Comment by Hainish** [comment]
@cschanaj in the latest Firefox versions this pref does not change the behavior as they've made signature checking manditory.

**Comment by cschanaj** [comment]
@Hainish okay. would you also review and merge this PR?

**Comment by Hainish** [comment]
@cschanaj yes but it'll have to wait until at least tomorrow

**Comment by cschanaj** [comment]
@Hainish never mind, but I would like this to be merged before the next release given the performance boost it will provides. thanks.

**Comment by Hainish** [comment]
This is great.  Efficiency can be further improved by utilizing a [radix trie](https://en.wikipedia.org/wiki/Radix_tree), indexed by the string reversal of the target (moc.elpmaxe) but this is contingent upon #11203

**Comment by cschanaj** [comment]
I agree a radix trie should be the best structure for the domains, but we will have to wait #11203 is completed. So I guess using Map could be a easy fix in the short term. :smile:

**Comment by koops76** [comment]
@Hainish Maybe use a bloom filter or a prefix list?

**Comment by cschanaj** [comment]
AFAIK, a bloom filter gives probabilistic results. And thus using a bloom filter might require a major rewrite.

**Comment by Hainish** [comment]
@koops76 bloom filters are only appropriate solutions when false positives are acceptable.  That's not true here - we don't want to DoS a site unintentionally!

**Comment by koops76** [comment]
@Hainish We can verify positives by hashes or simply a list of acceptable targets.

**Comment by cschanaj** [comment]
And a bloom filter give boolean results. The searching of applicable results is still inefficient.

**Comment by cschanaj** [comment]
@Hainish thanks! I will work on using JSON instead of XML for default.rulesets in a later PR.

**Pull request merged**